streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +39,678 @@
+                    changelogMetadata.bufferedRecords.addAll(records);
+            }
+
+            for (final TopicPartition partition: restoringChangelogs) {
+                // even if some partition do not have any accumulated data, we still trigger
+                // restoring since some changelog may not need to restore any at all, and the
+                // restore to end check needs to be executed still.
+                // TODO: we always try to restore as a batch when some records are accumulated, which may result in
+                //       small batches; this can be optimized in the future, e.g. wait longer for larger batches.
+                restoreChangelog(changelogs.get(partition));
             }
-        });
-
-        final Iterator<TopicPartition> iter = initializable.iterator();
-        while (iter.hasNext()) {
-            final TopicPartition topicPartition = iter.next();
-            final Long restoreOffset = restoreToOffsets.get(topicPartition);
-            final StateRestorer restorer = stateRestorers.get(topicPartition);
-
-            if (restorer.checkpoint() >= restoreOffset) {
-                restorer.setRestoredOffset(restorer.checkpoint());
-                iter.remove();
-                completedRestorers.add(topicPartition);
-            } else if (restoreOffset == 0) {
-                restorer.setRestoredOffset(0);
-                iter.remove();
-                completedRestorers.add(topicPartition);
+        }
+    }
+
+    /**
+     * restore a changelog with its buffered records if there's any; for active changelogs also check if
+     * it has completed the restoration and can transit to COMPLETED state and trigger restore callbacks
+     */
+    private void restoreChangelog(final ChangelogMetadata changelogMetadata) {
+        final ProcessorStateManager stateManager = changelogMetadata.stateManager;
+        final TopicPartition partition = changelogMetadata.changelogPartition;
+        final StateStoreMetadata storeMetadata = stateManager.storeMetadata(partition);
+        final String storeName = storeMetadata.stateStore.name();
+        final int numRecords = changelogMetadata.bufferedLimitIndex;
+
+        if (numRecords != 0) {
+            final List<ConsumerRecord<byte[], byte[]>> records = changelogMetadata.bufferedRecords.subList(0, numRecords);
+            stateManager.restore(partition, records);
+
+            // NOTE here we use removeRange of ArrayList in order to achieve efficiency with range shifting,
+            // otherwise one-at-a-time removal or addition would be very costly; if all records are restored
+            // then we can further optimize to save the array-shift but just set array elements to null
+            if (numRecords < changelogMetadata.bufferedRecords.size()) {
+                records.clear();
             } else {
-                restorer.setEndingOffset(restoreOffset);
+                changelogMetadata.bufferedRecords.clear();
```

Comment:
Is it really true that Java won't optimize `list.subList(0,list.size()).clear()` to `list.clear()`?

Reply:
From my read that's the case. For list.subList.clear() the subList#removeRange would be triggered, which does the shifting; for list.clear() we just set all values to `null` without copying.